File: Python3/Medium/Accepted/contiguous-array_7.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        
        max_len = 0
        sum = 0
        dict = {}
        for i in range(len(nums)):
            if nums[i] == 0:
                sum -= 1
            else:
                sum += 1
            if sum == 0:
                max_len = i + 1
            else:
                if sum in dict:
                    max_len = max(max_len, i - dict[sum])
                else:
                    dict[sum] = i
        return max_len

File: C++/Medium/Accepted/contiguous-array_7.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        
        int n = nums.size();
        int max_len = 0;
        int sum = 0;
        unordered_map<int, int> mp;
        mp[0] = -1;
        for (int i = 0; i < n; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if (mp.find(sum) != mp.end()) {
                max_len = max(max_len, i - mp[sum]);
            } else {
                mp[sum] = i;
            }
        }
        return max_len;
    }
};